hw/bsp/rp2040/family.cmake

```diff
@@ -1,77 +1,75 @@
-	${TOP}/src/portable/raspberrypi/${FAMILY}/rp2040_usb.c
-	${TOP}/src/portable/raspberrypi/${FAMILY}/dcd_rp2040.c
-	${TOP}/src/portable/raspberrypi/${FAMILY}/hcd_rp2040.c
-)
-
-target_sources(${PROJECT} PUBLIC
-  ${TOP}/hw/bsp/${FAMILY}/family.c
-  ${SRC_TINYUSB}
-)
-
-target_include_directories(${PROJECT} PUBLIC
-  ${TOP}/hw
-  ${TOP}/src
-  ${TOP}/hw/bsp/${FAMILY}/boards/${BOARD}
-)
-
-target_compile_definitions(${PROJECT} PUBLIC
-  CFG_TUSB_MCU=OPT_MCU_RP2040
-  PICO_RP2040_USB_DEVICE_ENUMERATION_FIX=1
-)
-
-if(DEFINED LOG)
-  target_compile_definitions(${PROJECT} PUBLIC CFG_TUSB_DEBUG=${LOG} )
-endif()
+cmake_minimum_required(VERSION 3.13)
+if (NOT TARGET _rp2040_family_inclusion_marker)
+	add_library(_rp2040_family_inclusion_marker INTERFACE)
 
-if(LOGGER STREQUAL "rtt")
-  target_compile_definitions(${PROJECT} PUBLIC
-    LOGGER_RTT
-    SEGGER_RTT_MODE_DEFAULT=SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL
-  )
+	# include basic family CMake functionality
+	set(FAMILY_MCUS RP2040)
 
-  target_sources(${PROJECT} PUBLIC
-    ${TOP}/lib/SEGGER_RTT/RTT/SEGGER_RTT.c
-  )
+	# add the SDK in case we are standalone tinyusb example (noop if already present)
+	include(${CMAKE_CURRENT_LIST_DIR}/pico_sdk_import.cmake)
 
-  target_include_directories(${PROJECT} PUBLIC
-    ${TOP}/lib/SEGGER_RTT/RTT
-  )
-endif()
+	include(${CMAKE_CURRENT_LIST_DIR}/../family.cmake)
+
+	# todo should we default to pico_sdk?
+	if (NOT BOARD)
+		message(FATAL_ERROR "BOARD must be specified")
+	endif()
+	include(${CMAKE_CURRENT_LIST_DIR}/boards/${BOARD}/board.cmake)
+
+	# TOP is absolute path to root directory of TinyUSB git repo
+	set(TOP "../../..")
+	get_filename_component(TOP "${TOP}" REALPATH)
+
+	if (NOT PICO_TINYUSB_PATH)
+		set(PICO_TINYUSB_PATH ${TOP})
+	endif()
+
+	# tinyusb_additions will hold our extra settings libraries
+	add_library(tinyusb_additions INTERFACE)
+
+	target_compile_definitions(tinyusb_additions INTERFACE
+		PICO_RP2040_USB_DEVICE_ENUMERATION_FIX=1
+		CFG_TUSB_MCU=OPT_MCU_RP2040 # this is already included in the SDK, but needed here for build_family.py grep
+	)
+
+	if(DEFINED LOG)
+	  target_compile_definitions(tinyusb_additions INTERFACE CFG_TUSB_DEBUG=${LOG} )
+	endif()
 
+	if(LOGGER STREQUAL "rtt")
+	  target_compile_definitions(tinyusb_additions INTERFACE
+		LOGGER_RTT
+		SEGGER_RTT_MODE_DEFAULT=SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL
+	  )
+
+	  target_sources(tinyusb_additions INTERFACE
+		${TOP}/lib/SEGGER_RTT/RTT/SEGGER_RTT.c
+	  )
+
+	  target_include_directories(tinyusb_additions INTERFACE
+		${TOP}/lib/SEGGER_RTT/RTT
+	  )
+	endif()
+
+	function(family_configure_target TARGET)
+		pico_add_extra_outputs(${TARGET})
+		pico_enable_stdio_uart(${TARGET} 1)
+		target_link_libraries(${TARGET} PUBLIC pico_stdlib pico_bootsel_via_double_reset tinyusb_board tinyusb_additions)
+	endfunction()
+
+	function(family_configure_device_example TARGET)
+		family_configure_target(${TARGET})
+		target_link_libraries(${TARGET} PUBLIC pico_stdlib tinyusb_device)
```

Comment:
yes that is part of my plan, i just wasn't 100% sure i wanted to chew that off for the SDK1.2.0 release which I'd like to do asap. I'll see how today looks. It's all a bit of a dance as we support both tinyusb->sdk and sdk->tinyusb and migrate code which clearly belongs in tinyusb from sdk

Reply:
no problem, take your time :)

Reply:
yup, there is already a comment on the other side :-)

https://github.com/raspberrypi/pico-sdk/blob/707d1c20cba2eb0ae791ab469520aa24df44a6e6/src/rp2_common/tinyusb/CMakeLists.txt#L20

Reply:
it would be pretty trivial, but we have some stuff to make sure we don't mix the CDC stdio stuff with explicit use of tinyusb, which was the part i wanted tot tackle in the next release.

Reply:
Except for this, I think the PR is ready for merge. In general, pico-sdk and tinyusb should not reply too much on other cmake defines/variable, that would make it easier to change in both sides without affecting others. I could imagine to do quite a bit of work with cmake when migrating all other ports to cmake in the future (just need to fix pending bugs first).

Reply:
>  I think the example in this repo should always built with relative path to this code

I've lost track of the context for this ; what _relative path_ of what _code_?

Reply:
Note also, I was not planning to futher disambiguate the builds for this SDK release (which is super imminent), so I'd prefer to merge this as is so that upstream tinyusb is compatible with (and can be merged into the raspberry-pi version by users wanting new stuff) until we fix in a point release. You've already released 0.10.0 anyway

Reply:
> > I think the example in this repo should always built with relative path to this code
> 
> I've lost track of the context for this ; what _relative path_ of what _code_?

I mean path to `/src` from the example. 

> Note also, I was not planning to futher disambiguate the builds for this SDK release (which is super imminent), so I'd prefer to merge this as is so that upstream tinyusb is compatible with (and can be merged into the raspberry-pi version by users wanting new stuff) until we fix in a point release. You've already released 0.10.0 anyway

the list of sources (tinyusb_device/ host) must be specified within upstream for normal developing to continue without modifying the sdk or wait for follow up PR. It can be hard corded (and ugly) like I did with esp32s2 https://github.com/hathach/tinyusb/blob/master/examples/device/cdc_msc_freertos/src/CMakeLists.txt#L18 
 but is modifiable within this repo. This is unfortunately a must before merging.

Reply:
ok i have moved the definitions... is it possible to do a 0.10.1 release with this (and including the other rp2040 fix which is already in master) as we can then point the SDK 1.2.0 (planning to release tomorrow) directly at this repo rather than the raspberry pi copy

Reply:
@kilograham could you fix the ci build, it seems to failed with rp2040

Reply:
@kilograham seems like you have issue with ci when moved to this. I take this requirement back. Please revert those moving commits if you still want it. I will merge the PR as it is previously (cmake source list in sdk), then make a release to 0.10.1.

I could follow it up to change later on. I need to update the cmake build anyway.  you are probably working too much already on sdk 1.2.0. What do you say  ?

Reply:
almost fixed the issues - been busy with other stuff (which were unrelated - more to do with pico-examples maybe not having tinyusb/lwip sub module initialized in which case we shouldn't build lwip examples).

If we keep the lists separate, then there is no guarantee that things are in sync... so it is a bit of 6 of one a dozen of the other. You aren't going to unilaritarily break the SDK build any more than if you were to commit a bad tinyusb other change. We only pick it up a specific commit in our tinyusb submodule

Reply:
@kilograham great, thanks, I am bit worried that you are overworked. Sound good, I will request your review if making changes to cmake, also test it out within pico-sdk as well.